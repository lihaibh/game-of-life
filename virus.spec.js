var sinon = require('sinon');
var { Membrane } = require('./membrane');
var { Grid, cellReducer } = require('./grid');
var { Virus, infectedCellBehavior } = require('./virus');
var { Observable, timer } = require('rxjs');

var should = require('should');
require('rxjs/add/operator/takeUntil');
require('rxjs/add/operator/finally');

// Make sure the virus functionality is ok
describe('Virus', () => {
    it('should infect a membrane by changing it\'s cells behavior when the virus has the right conditions to spread', () => {
        var width = 3;
        var height = 2;

        // create a membrane to infect
        var grid = Grid.create(width, height, "1 0 0 1 1 0");
        var timeout = timer(5000);
        var maxGenerations = 6;
        var finishCondition = (grid) => grid.generations == maxGenerations;

        var spyError = sinon.spy((err) => console.log(err));
        var spyComplete = sinon.spy();
        var spyCheckExpectedState = sinon.spy((state) => {
            expect(state).toMatchSnapshot();
        });

        var membrane = new Membrane(grid, finishCondition);
        var stateUnderTest$ = membrane.state$().takeUntil(timeout);

        // as the virus infect the membrane we should expect the reduce function of the cells to be replaced
        stateUnderTest$.subscribe(spyCheckExpectedState, spyError, spyComplete);

        stateUnderTest$.finally(() => {
            spyError.callCount.should.be.eql(0);
            spyComplete.callCount.should.be.eql(1);

            // the first call is for getting the initial state and the others is for new generations
            spyCheckExpectedState.callCount.should.be.eql(7);
        }).subscribe(() => { });

        // create the virus under test
        var virus = new Virus((state) => state.generations == 5);
        virus.infect(membrane);

        membrane.run();
    });

    it('should not infect a membrane if it doesnt have the right conditions', () => {
        var width = 3;
        var height = 2;

        // create a membrane to infect
        var grid = Grid.create(width, height, "1 0 0 1 1 0");
        var timeout = timer(5000);
        var maxGenerations = 6;
        var finishCondition = (grid) => grid.generations == maxGenerations;

        var spyError = sinon.spy((err) => console.log(err));
        var spyComplete = sinon.spy();
        var spyCheckExpectedState = sinon.spy((state) => {
            expect(state).toMatchSnapshot();
        });

        var membrane = new Membrane(grid, finishCondition);
        var stateUnderTest$ = membrane.state$().takeUntil(timeout);

        // as the virus infect the membrane we should expect the reduce function of the cells to be replaced
        stateUnderTest$.subscribe(spyCheckExpectedState, spyError, spyComplete);

        stateUnderTest$.finally(() => {
            spyError.callCount.should.be.eql(0);
            spyComplete.callCount.should.be.eql(1);

            // the first call is for getting the initial state and the others is for new generations
            spyCheckExpectedState.callCount.should.be.eql(7);
        }).subscribe(() => { });

        // create the virus under test - it will never have the right conditions to spread
        var virus = new Virus((state) => false);
        virus.infect(membrane);

        membrane.run();
    });
});