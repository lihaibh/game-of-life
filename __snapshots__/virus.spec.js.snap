// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Virus should infect a membrane by changing it's cells behavior when the virus has the right conditions to spread 1`] = `
Object {
  "generations": 0,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should infect a membrane by changing it's cells behavior when the virus has the right conditions to spread 2`] = `
Object {
  "generations": 1,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should infect a membrane by changing it's cells behavior when the virus has the right conditions to spread 3`] = `
Object {
  "generations": 2,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should infect a membrane by changing it's cells behavior when the virus has the right conditions to spread 4`] = `
Object {
  "generations": 3,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should infect a membrane by changing it's cells behavior when the virus has the right conditions to spread 5`] = `
Object {
  "generations": 4,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should infect a membrane by changing it's cells behavior when the virus has the right conditions to spread 6`] = `
Object {
  "generations": 5,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should infect a membrane by changing it's cells behavior when the virus has the right conditions to spread 7`] = `
Object {
  "generations": 6,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should not infect a membrane if it doesnt have the right conditions 1`] = `
Object {
  "generations": 0,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should not infect a membrane if it doesnt have the right conditions 2`] = `
Object {
  "generations": 1,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should not infect a membrane if it doesnt have the right conditions 3`] = `
Object {
  "generations": 2,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should not infect a membrane if it doesnt have the right conditions 4`] = `
Object {
  "generations": 3,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should not infect a membrane if it doesnt have the right conditions 5`] = `
Object {
  "generations": 4,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should not infect a membrane if it doesnt have the right conditions 6`] = `
Object {
  "generations": 5,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;

exports[`Virus should not infect a membrane if it doesnt have the right conditions 7`] = `
Object {
  "generations": 6,
  "grid": Grid {
    "height": 2,
    "matrix": Array [
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
      ],
      Array [
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": false,
          },
        },
        Object {
          "reducer": [Function],
          "state": Object {
            "alive": true,
          },
        },
      ],
    ],
    "width": 3,
  },
}
`;
